backend/src/Gitlab/Service/GitlabMRService.php

```diff
@@ -0,0 +1,295 @@
+        $diffs = $client->request('GET', "merge_requests/{$gitlabMRId}/diffs")->toArray();
+        $response['commits'] = $client->request('GET', "merge_requests/{$gitlabMRId}/commits")->toArray();
+
+        $response['files'] = array_map(fn($file) => [
+            'patch' => $file['diff'] ?? null,
+            'filename' => $file['new_path'] ?? $file['old_path'] ?? null,
+            'status' => $this->getDiffFileStatus($file),
+        ], $diffs);
+        // filter out non-text files
+        $response['files'] = array_filter($response['files'], static fn($file) => $file['patch'] !== null);
+
+        return GitlabMRResponseDTO::fromAPIResponse($response);
+    }
+
+    private function getGitlabProjectClient(ProjectAgentConnection $connection, string $repoURL = ''): HttpClientInterface
+    {
+        $repoURL = $connection->getConfigValue('repository') ?: $repoURL;
+        if (!$repoURL) {
+            throw new \RuntimeException('Repository URL not found');
+        }
+        if (!str_contains($repoURL, 'https://gitlab.com/')) {
+            $parsedURL = parse_url($repoURL);
+            if (!$parsedURL) {
+                throw new \RuntimeException('Invalid repository URL');
+            }
+            $baseURL = $parsedURL['scheme'] . '://';
+            if (isset($parsedURL['user'])) {
+                $baseURL .= $parsedURL['user'];
+            }
+            if (isset($parsedURL['pass'])) {
+                $baseURL .= ':' . $parsedURL['pass'];
+            }
+            if (isset($parsedURL['user'])) {
+                $baseURL .= '@';
+            }
+            $baseURL .= $parsedURL['host'];
+            if (isset($parsedURL['port'])) {
+                $baseURL .= ':' . $parsedURL['port'];
+            }
+            $apiURL = $baseURL . '/api/v4';
+        } else {
+            $apiURL = 'https://gitlab.com/api/v4';
+        }
+        $apiURL .= '/projects/' . $this->getProjectPath($repoURL) . '/';
+
+        // You can also use personal, project, or group access tokens with OAuth-compliant headers:
+        $token = $connection->getConfigValue('gitlabToken') ?: $connection->agent->accessKey;
+
+        return $this->httpClient->withOptions([
+            'base_uri' => $apiURL,
+            'headers' => [
+                'Authorization' => 'Bearer ' . $token,
+            ],
+        ]);
+    }
+
+    private function getProjectPath(string $repoURL): string
+    {
+        $parsedURL = parse_url($repoURL, PHP_URL_PATH);
+        if (!$parsedURL) {
+            throw new \RuntimeException('Invalid repository URL');
+        }
+
+        return rawurlencode(trim($parsedURL, '/'));
+    }
+
+    private function getDiffFileStatus(array $file): string
+    {
+        if (!empty($file['new_file'])) {
+            return 'created';
+        }
+        if (!empty($file['deleted_file'])) {
+            return 'deleted';
+        }
+
+        return 'modified';
+    }
+}
```

Comment:
BOT4:
The file does not end with a newline. According to POSIX standards and general coding conventions, text files should end with a newline character.